### PR TITLE
Remove "big data" from the database

### PR DESCRIPTION
**Pull request by inducer**
Before this change, Relate stores uploaded files in the database, which makes DB dumps (mainly for backups) unwieldy and unpleasant to work with. It likely also needlessly slows down the database. The large (20-ish GB) database dumps are one main reason why running Relate has become an unpleasant chore for me. This is intended to help address that.

With this change, new submissions will be stored in the file system (or, to be technically correct, a configurable Django `Storage`). Old submissions can be migrated over time with a new management command that this introduces.

I'd be grateful for reviews and testing here, particularly since there's a very real possibility of data loss in case there are issues.

cc @joshuata

### CODE REVIEW

**Comment by joshuata** [comment]
One potential issue I see is one we dealt with - serving large uploaded files securely. It's not too much of a problem at < 2 MB, but any larger and you are blocking the django thread for a long time. We dealt with this by using [django-sendfile2](https://pypi.org/project/django-sendfile2/) to serve the files from nginx/apache while still using the application permissions layer.

Its worth noting that django-sendfile2 is customizeable using settings, so you can serve using the django webserver by default while still providing a graceful upgrade path for users who are concerned about responsiveness and security for larger files in exchange for managing another webserver

This may be more than you want to deal with right now, but it also might be worth looking at.

**Comment by dzhuang** [comment]
That's a good point!

**Comment by dzhuang** [comment]
Will you consider [`django-sendfile`](https://github.com/johnsensible/django-sendfile) for the storage backend? It will be good for serving large files (especially which are not able to be previews with browser) while allow add permssions to the file.

**Comment by dzhuang** [comment]
@joshuata Just notices your comments. I happen to have the same idea as you. I didn't know there's a version 2.

**Comment by joshuata** [comment]
Yeah. Version 2 is python 3 and django 2.1+ compatible. The old version stalled out about 3 years ago now and hasn't had any updates since

**Comment by inducer** [comment]
> One potential issue I see is one we dealt with - serving large uploaded files securely. It's not too much of a problem at < 2 MB, but any larger and you are blocking the django thread for a long time.

I agree! I think we'll want work along these lines at some point, however step 1 perhaps would be to not serve submissions as a `data:` URI. :P (I'm honestly still somewhat surprised that we haven't been more seriously been bitten by that.) There's a bit of engineering in that, too: Pages need to be able to define their own URL endpoints then. Once that's done, starting to use `django-sendfile2` would be within reach. But I think it's legitimately out of scope for this PR. Could you open an issue pointing back to this discussion?

File: course/page/upload.py [hunk 1]
```diff
@@ -209,20 +209,61 @@ def markup_body_for_title(self):
...
+            if page_context.flow_session.participation is not None:
+                username = page_context.flow_session.participation.user.username
+            if page_context.flow_session.flow_id:
+                flow_id = page_context.flow_session.flow_id
+
+        return (f"{page_context.course.identifier}/"
+                f"{flow_id}/"
+                f"{self.page_desc.id}/"
+                f"{username}"
+                f"{ext}")
```
[the remark this answers is not in the document]
Comment (inducer) [reply course/page/upload.py:232]: Yep, exactly. :)

**Comment by inducer** [comment]
Thanks for taking a look!

**Comment by inducer** [comment]
This is nominally done. Testing on a backup copy of the Illinois data set before merging, will report results.

**Comment by inducer** [comment]
The previous version created giant in-memory data structures. This is now fixed, and seems to be running OK on the Illinois data (still going, will report results).

**Comment by inducer** [comment]
This latest version extracted 38 GB of data from the database, and the compressed database dump is now 907 MB. (23GB previously). NB: I've also run `manage.py clearsessions` between then and now, for another ~200 MB of shrinkage.

Creating the (`gzip`-compressed) dump out of an (vacuumed) database now takes ~3m (vs. close to an hour previously)

Here are the biggest relation sizes reported by
```sql
SELECT nspname || '.' || relname AS "relation",
    pg_size_pretty(pg_relation_size(C.oid)) AS "size"
  FROM pg_class C
  LEFT JOIN pg_namespace N ON (N.oid = C.relnamespace)
  WHERE nspname NOT IN ('pg_catalog', 'information_schema')
  ORDER BY pg_relation_size(C.oid) DESC
  LIMIT 20;
```
after this cleanup and `vacuum full`:
```
 public.course_flowpagevisitgrade                                | 4693 MB
 public.course_flowpagevisit                                     | 2655 MB
 public.course_flowpagevisit_page_data_id_2561aa392d731f21_uniq  | 755 MB
 public.course_flowpagevisit_e8701ad4                            | 538 MB
 public.course_flowpagevisit_2bd7de0f                            | 538 MB
 public.course_flowpagevisit_5ba34997                            | 538 MB
 public.course_flowpagevisit_fdd1ff75                            | 538 MB
 public.course_flowpagevisit_pkey                                | 538 MB
 public.course_flowpagevisit_93f2a5a7                            | 538 MB
 public.course_flowpagedata                                      | 527 MB
 public.course_flowpagevisitgrade_visit_id_6860fb32b87ae4cc_uniq | 312 MB
 public.course_flowpagebulkfeedback                              | 227 MB
 public.course_flowpagevisitgrade_768334c3                       | 222 MB
 public.course_flowpagevisitgrade_ea367887                       | 222 MB
 public.course_flowpagevisitgrade_pkey                           | 222 MB
 public.course_flowpagevisitgrade_f9d4d9d8                       | 222 MB
 public.course_flowsession                                       | 120 MB
 public.course_flowpagedata_5ba34997                             | 95 MB
 public.course_flowpagedata_pkey                                 | 95 MB
 public.course_flowpagebulkfeedback_page_data_id_key             | 66 MB
```

**Comment by inducer** [comment]
@dzhuang, if you'd like to take a final look, let me know. Otherwise I'll probably merge this this weekend.

**Comment by dzhuang** [comment]
Sorry, I failed to check this email notification on time (buried), but I think it's ok for me. Thanks.

**Pull request merged**